### ISSUE TITLE
Missing `warn(unused_mut)` in `for` loop with nightly-2017-12-28

### ISSUE THREAD

**Issue by uHOOCCOOHu** [rust-lang/rust#47070]
All these warnings are triggered in stable-1.22.1.
```rust
fn main() {
    // let mut t = 1;   // triggered
    for mut i in 0..5 { // not triggered
        print!("{} ", i);
    }
    for _ in 0..5 {
        let mut x = 1;  // not triggered
        print!("{} ", x);
    }
    for _ in 0..5 {
        if true {
            match 2 {
                mut x => print!("{} ", x), // not triggered
            }
        }
    }
}
```

**Comment by fabric-and-ink** [comment]
This works on current stable. [Link to Playground.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81a5ca3c92d4f9953517007a3a434fcb)
